### PR TITLE
Add Hierarchical Stats example

### PR DESCRIPTION
**Pull request by apatole**
### Description

This patch adds hierarchical stats example and the related notebooks to try it out.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [x] In-line docstrings updated.
- [x] Documentation updated.

### CODE REVIEW

**Comment by apatole** [comment]
Hi @chesterxgchen @yanchengnv Please help review PR to add hierarchical statistics example, thanks.

File: examples/advanced/federated-statistics/README.md [hunk 1]
```diff
@@ -89,6 +89,40 @@ via the MONAI FLARE integration.
 load and cache the externally calculated statistics. The server side controller will pass the target metrics to pre_run method 
 so it can be used to load the statistics. 
 
+### Hierarchical Stats example
```
[the remark this answers is not in the document]
Comment (apatole) [reply examples/advanced/federated-statistics/README.md:92]: Done, thanks.

**Comment by apatole** [comment]
Thanks @chesterxgchen for the Review!
I have incorporated review comments and uploaded a new patch set.

File: examples/advanced/federated-statistics/hierarchical_stats.ipynb [hunk 2]
```diff
@@ -0,0 +1,236 @@
+{
+ "cells": [
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats.ipynb:2]: suggest moving this notebook file inside the "hierarchical_stats" folder.

File: examples/advanced/federated-statistics/README.md [hunk 3]
```diff
@@ -89,6 +89,72 @@ via the MONAI FLARE integration.
...
+
+Here is an example of the generated hierarchical statistics for the students data from different universities.
+
+Expandable global and level wise stats
+
+<img src="hierarchical_stats/demo/expandable_stats.png" alt="Expandable Hierarchical stats" width="800">
+
+Expanding `Global` will show visualization of global statistics similar to the following
+
+<img src="hierarchical_stats/demo/global_tats.png" alt="Global stats" width="800">
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/README.md:138]: typo on file name global_tats.png -> global_stats.png

File: examples/advanced/federated-statistics/hierarchical_stats/README.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+```
+You should be able to get the visualization similar to the followings
+
+Expandable global and level wise stats, clicking on each will shows statistics for that particular level.
+
+<img src="demo/expandable_stats.png" alt="Expandable Hierarchical stats" width="400">
+
+Expanding `Global` will show visualization of global statistics similar to the following
+
+<img src="demo/global_tats.png" alt="Global stats" width="400">
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats/README.md:86]: typo on file name global_tats.png -> global_stats.png

File: examples/advanced/federated-statistics/hierarchical_stats/demo/global_tats.png
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats/demo/global_tats.png:1]: typo on file name global_tats.png -> global_stats.png

File: examples/advanced/federated-statistics/hierarchical_stats.ipynb [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "52d2a52f-8a8d-45ef-8a50-ddbb4ed2f2c6",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "We are not quite done yet. What if you prefer to use python API instead CLI to run jobs. Lets do that in this section"
+   ]
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats.ipynb:201]: this line confuses me a bit, what "section" are we referring to here?

File: examples/advanced/federated-statistics/hierarchical_stats.ipynb [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3cd042db-6ce0-4e37-bcbe-d96051e4d164",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Visualization\n",
+    "We can visualize the results easly via the visualizaiton notebook. Before we do that, we need to copy the data to the notebook directory \n"
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats.ipynb:170]: typo: visualizaiton -> visualization

File: examples/advanced/federated-statistics/hierarchical_stats/README.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+          "stddev": {},
+          "histogram": {
+            "*": { "bins": 2, "range": [0, 1] },
+            "Percentage": {"bins": 4, "range": [0.0, 100.00] }
+          }
+        },
+        "writer_id": "stats_writer",
+        "enable_pre_run_task": false,
+        "precision": 4,
+        "hierarchy_config": "hierarchy_config.json"
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats/README.md:148]: could you add some explanation on any requirements/structure of this "hierarchy_config"?
Does it have to be a JSON file? Is there any key that must exist etc?

File: examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    "#### dependencies\n",
+    "\n",
+    "To run examples, you might need to install the dependencies\n",
+    "* ipywidgets\n",
+    "* numpy\n",
+    "* pandas\n",
+    "* matplotlib\n",
+    "* jupyter\n",
+    "\n",
+    "These are captured in the requirements.txt"
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:25]: can we do pip install ../requirements.txt as you did in the other notebook?

File: examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "1b0f21fd",
+   "metadata": {},
+   "source": [
+    "### Overall Statistics\n",
+    "\n",
+    "In this section, we demonstrate how to visualize hierarchical stats from the tabular data."
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:80]: suggest: "In this section, we demonstrate how to visualize hierarchical stats from the tabular data using Jupyter Widgets."

File: examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+   "source": [
+    "### Histogram Visualization\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3e86860e",
+   "metadata": {},
+   "source": [
+    "The following code demonstrates how the hierarchical histograms can be displayed display for for features like `percentage`."
```
Comment (YuanTingHsieh) [inline examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:146]: suggest "The following code demonstrates how the hierarchical histograms can be displayed using Jupyter Widgets for features like `percentage`."

File: examples/advanced/federated-statistics/hierarchical_stats.ipynb [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "26cb3afa",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "# Federated Hierarchical Statistics \n",
+    "\n",
+    "In this example, we will show how to generate federated Hierarchical statistics for data that can be represented as Pandas Data Frame.\n",
```
[the remark this answers is not in the document]
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats.ipynb:12]: Done, thanks.

File: examples/advanced/federated-statistics/hierarchical_stats/utils/prepare_data.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import argparse
+import csv
+import os
+import random
+import sys
```
[the remark this answers is not in the document]
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats/utils/prepare_data.py:18]: Done, thanks.

File: examples/advanced/federated-statistics/hierarchical_stats.ipynb (hunk 2 above)
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats.ipynb:2]: Done, thanks.

File: examples/advanced/federated-statistics/hierarchical_stats/README.md (hunk 4 above)
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats/README.md:86]: Done, thanks.

File: examples/advanced/federated-statistics/hierarchical_stats.ipynb (hunk 6 above)
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats.ipynb:170]: Done, thanks.

File: examples/advanced/federated-statistics/hierarchical_stats/README.md (hunk 7 above)
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats/README.md:148]: Done, thanks. Sorry I missed this important part earlier.

File: examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb (hunk 9 above)
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:80]: Done, thanks.

File: examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb (hunk 10 above)
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:146]: Done, thanks.

**Comment by apatole** [comment]
Many thanks @yanchengnv for the review, I have uploaded updated patch incorporating review comments.

Couple of  additional changes:

- Updated `examples/advanced/federated-statistics/README.md` with the link to hierarchical stats.
- Fixed existing broken link in examples/advanced/federated-statistics/README.md

File: examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb (hunk 8 above)
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats/demo/visualization.ipynb:25]: Done, thanks.

File: examples/advanced/federated-statistics/hierarchical_stats.ipynb (hunk 5 above)
Comment (apatole) [reply examples/advanced/federated-statistics/hierarchical_stats.ipynb]: Actually that line got copied from the other notebook I referred. Removed it now as I don't see it relevant here. Thanks.

**Review by YuanTingHsieh — APPROVED** [review]
thanks, LGTM

**Comment by apatole** [comment]
Thanks @SYangster for triggering a build and helping with `blossom-ci`, all checks now passed.

Looks like someone with the write access need to merge it.

**Pull request merged**